Consent fixes

alxdarksage opened the pull request:
BRIDGE-2182

1) always indicate that the date is in GMT, because it always is (we most likely don't know the user's time zone at this point);

2) If the account has a verified email, use it, else a verified phone, use it, else use an external ID if it exists, otherwise, just don't show an identifier on the final consent document

3) This uncovered a bug related to creating a consent record with a credential before it is verified; I've moved ITP processing from sign up to sign in to avoid this. There's more places it needs to be inserted, but works the same way... before throwing a consent exception, we check to see if there's a pending consent and process that first, then continue.

Also filed this issue based on functionality I uncovered working on this: https://sagebionetworks.jira.com/browse/BRIDGE-2191

alxdarksage committed:
Move ITP processing from sign up to sign in, to ensure a credential is verified for the consent.

DwayneJengSage commented on app/org/sagebionetworks/bridge/services/AuthenticationService.java:
```diff
@@ -171,6 +171,10 @@ public UserSession signIn(Study study, CriteriaContext context, SignIn signIn) t
 
+        if (intentService.registerIntentToParticipate(study, account)) {
+            account = accountDao.getAccount(AccountId.forId(study.getIdentifier(), account.getId()));    
+        }
```
It looks like this is non-trivial and it will be executed on every sign-in, which could be a lot. Would it be better to move this to verify channel, which is most likely only executed once?

(Similarly with channelSignIn())

alxdarksage replied:
Hmm I hadn't thought of that. Yes that's a good idea. I'll look at moving it to the verification step. We just need to be sure we verify through that method if people sign in through email or phone, without explicitly verifying.

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java:
```diff
@@ -151,7 +153,7 @@
     private Study study;
 
-    private Account account;
+    private GenericAccount account;
```
We should avoid direct references to GenericAccount, since this bypasses our abstraction, and makes it more difficult to merge HibernateAccount and GenericAccount in the future (if that's the path we're planning on going). If there's something we need in GenericAccount outside of the AccountDao, we should probably move it to the interface.

alxdarksage replied:
Added setId and setStudyId (which does not match getStudyIdentifier, but I'm not going to change that in this PR).

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/ConsentPdfTest.java:
```diff
@@ -125,13 +123,60 @@ public void legacyDocWithInvalidSig() throws Exception {
+        consentPdf.getFormattedConsentDocument();
+    }
+    
+    public void dateFormattedCorrectly() throws Exception {
```
Missing Test annotation?

alxdarksage replied:
Added

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/ConsentPdfTest.java:
```diff
@@ -125,13 +123,60 @@ public void legacyDocWithInvalidSig() throws Exception {
+                NEW_DOCUMENT_FRAGMENT, consentBodyTemplate);
+        String output = consentPdf.getFormattedConsentDocument();
+        
+        assertTrue("Contains formatted date", output.contains(ConsentPdf.FORMATTER.print(DateTime.now()) + " (GMT)"));
```
It would be better to use DateUtils to mock now and verify the output against a known string. Otherwise, if ConsentPdf.FORMATTER formats the date incorrectly, this test will never catch it.

alxdarksage replied:
Changed to use a fixed point in time

DwayneJengSage commented on app/org/sagebionetworks/bridge/services/IntentService.java:
```diff
@@ -133,6 +140,27 @@ public void registerIntentToParticipate(Study study, StudyParticipant participan
         }
     }
     
+    public boolean registerIntentToParticipate(Study study, Account account) {
```
The above registerIntentToParticipate() appears to no longer be used, and this appears to be nearly identical. Instead of creating an overload, would it be better to instead change the old one?

alxdarksage replied:
This was a mistake on my part, this is mid-refactor. The participant method can go away, removed and altered tests that were testing it to test the other method

DwayneJengSage commented on app/org/sagebionetworks/bridge/services/IntentService.java:
```diff
@@ -133,6 +140,27 @@ public void registerIntentToParticipate(Study study, StudyParticipant participan
+                consentService.consentToResearch(study, subpop.getGuid(), participant, 
+                        intent.getConsentSignature(), intent.getScope(), true);
+                cacheProvider.removeObject(cacheKey);
+                return true;
```
Could an ITP represent multiple consents? If so, instead of returning true and short-circuiting, would it be better to save the result somewhere and return it later?

alxdarksage replied:
Theoretically it could. Changing this, also keeping track as to whether or not we've loaded the study participant.

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/AuthenticationServiceTest.java:
```diff
@@ -161,7 +161,7 @@ public void signUpWithIntentToParticipate() throws Exception {
             study.setAutoVerificationEmailSuppressed(true);
             study.setAutoVerificationPhoneSuppressed(true);
-            holder = authService.signUp(study, participant, true);
+            holder = authService.signUp(study, participant);
```
Does this test still succeed? If so, how?

alxdarksage replied:
Yes because it's really looking to see what happens when you sign in, and we've moved ITP consent processing from sign up to sign in (which has no flag). However, I just moved it to the verification step, so this test has to be changed again. However it's still looking to verify that sign in after this does not throw an exception.

alxdarksage committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/BridgePF into consent-fixes

alxdarksage committed:
Updates after code review.

alxdarksage committed:
Changing back to sign in based ITP processing, but only do it if the user is not consented.

alxdarksage committed:
remove comment

alxdarksage commented:
Updated. Moved ITP processing to consent rather than sign up, after authentication and only if the user is not consented (and would receive a 412 response). The associated integration test passes, but only by enabling the account without enabling the phone number, then signing in with a password. This is how our system works though I'm not sure this is really what we want. If we change this, we'll need some way for a caller to verify phone/email for testing... maybe an administrative UI to manipulate these for testing.

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java:
```diff
@@ -735,8 +738,107 @@ public void creatingExternalIdOnlyAccountSucceedsIfIdsManaged() {
+    @Test
+    public void signInWithIntentToParticipate() throws Exception {
+        account.setId(USER_ID);
+        Account consentedAccount = new GenericAccount();
```
We should create a Account.create() so we don't have references to GenericAccount everywhere. (Here and several other places in this PR.)

alxdarksage replied:
Done.

DwayneJengSage commented on app/org/sagebionetworks/bridge/services/IntentService.java:
```diff
@@ -114,23 +121,29 @@ public void submitIntentToParticipate(IntentToParticipate intent) {
             if (intent != null) {
+                if (participant == null) {
+                    participant = participantService.getParticipant(study, account.getId(), true);
+                }
```
Is getParticipant() a particularly expensive call? If it's not, it might be simpler to call getParticipant() before the loop, instead of having these extra if statements.

alxdarksage replied:
Yes, it's expensive here because we're loading all the consent records and we do use them in the consentToResearch call. I would be careful not to load it unless it's really necessary. Admittedly, the chance of their being two ITP consents (or more) is low and this call is a one-time deal, so we could remove this check.

alxdarksage committed:
Updates after review.
